parlai/nn/r3f.py

```diff
@@ -0,0 +1,190 @@
+
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        # Find embedding values and store locally so we don't have to find them each turn
+        self.r3f_embeddings = {}
+        self.is_generative_model = True  # assume true until we find otherwise
+        self._find_embeddings(self.model)
+
+    def compute_loss(self, batch, return_output=False):
+        if not hasattr(self, 'r3f_lambda'):
+            self.set_r3f_settings_from_opts(self.opt)
+        loss, standard_output = super().compute_loss(batch, True)
+        with R3FNoiseEmbeddingContextManager(self, self.model) as r3f:
+            noised_result = self.model(*self._model_input(batch), ys=batch.label_vec)
+            noised_scores, _, *_ = noised_result
+            standard_scores, _, *_ = standard_output
+            r3f_loss = r3f._calculate_symm_kl(noised_scores, standard_scores)
+            # get average loss per token correctly
+            notnull = batch.label_vec.ne(self.NULL_IDX)
+            target_tokens = notnull.long().sum(dim=-1)
+            r3f_loss /= target_tokens.sum()
+            # add in loss
+            loss += self.r3f_lambda * r3f_loss
+        if return_output:
+            return (loss, standard_output)
+        else:
+            return loss
+
+    def _calculate_symm_kl(self, noised_logits, input_logits):
+        return F.kl_div(
+            F.log_softmax(noised_logits, dim=-1, dtype=torch.float32),
+            F.softmax(input_logits, dim=-1, dtype=torch.float32),
+            None,
+            None,
+            "sum",
+        ) + F.kl_div(
+            F.log_softmax(input_logits, dim=-1, dtype=torch.float32),
+            F.softmax(noised_logits, dim=-1, dtype=torch.float32),
+            None,
+            None,
+            "sum",
+        )
+
+    def _find_embeddings(self, module):
+        self.is_generative_model ^= not re.search(".*GPT.*", module.__class__.__name__)
+
+        for name, layer in module.named_modules():
+            self.is_generative_model ^= not re.search(".*GPT.*", name)
+            if self.noise_encoder and (re.search(f"encoder.*norm_embeddings$", name)):
+                self.r3f_embeddings["encoder"] = layer
+            if self.noise_decoder and (
+                re.search(f"decoder.*norm_embeddings$", name)
+                or re.search("decoder.*wte$", name)  # cause GPT2 does it differently
+            ):
+                self.r3f_embeddings["decoder"] = layer
+        if (self.noise_encoder is True and "encoder" not in self.r3f_embeddings) or (
+            self.noise_decoder is True and "decoder" not in self.r3f_embeddings
+        ):
+            raise RuntimeError(
+                """
+                R3F: Embedding to add noise to not found in model. (Does the part, encoder/decoder, exist in the model you are running? Is the name for the embedding accounted for in `_find_embeddings()`?)
```

Comment:
wrap please.

Reply:
```python
(
   "This is a wrapped "
   "single line string "
   "in python."
)
```

Reply:
I don't get why the autoformatter doesn't just... do this. 

Maybe I'll take a look at it at some point.